### PR TITLE
Do not expect a defining root op when trying to undo complex pattern

### PR DESCRIPTION
**Pull request by jeanPerier**
The previous code was doing an unconditional usage of getDefiningOp result on the root of the two fir.complex insertOps to check that it was an undefOp.
This is wrong because it is not guaranteed that the root is not a block or function argument, especially in case of later mismatch.

This actually caused a bug #806 from Fortran code because some MLIR optimization pass combined some code operating on complexes into a new block while leaving the root undefOps in the parent op and passing the root as block argument (I think it is weird, but I have no idea what exactly is doing this, and this specific rewrite is hard to trigger, nearly any modification to #806 code stop MLIR from doing this rewrite).  In the case of #806, the pattern for which a match is attempted is anyway not a complex addition or subtraction, but it cannot be ruled out that similar control flow rewrites could happen on valid complex add/sub patterns.

Given the pattern rewrite is already checking the result is coming from two insertValueOp on complex that insert in 0 and 1, it actually does not matter to check that the root is an undefOp. So simply remove this check.

Add a test that operations can still be grouped even if the UndefOp is not visible. Also add fir tests for these complex pattern rewrites since there was none yet (there was only tests from Fortran).

### CODE REVIEW

File: flang/test/Fir/undo-complex-pattern.fir [hunk 1]
```diff
@@ -0,0 +1,96 @@
...
+  %imag1 = fir.extract_value %z1, %c0 : (!fir.complex<8>, index) -> f64
+  %real2 = fir.extract_value %z2, %c0 : (!fir.complex<8>, index) -> f64
+  %imag2 = fir.extract_value %z2, %c1 : (!fir.complex<8>, index) -> f64
+
+  // CHECK-LABEL: subf
+  // CHECK-LABEL: subf
+  %real = subf %real1, %real2 : f64
+  %imag = subf %imag1, %imag2 : f64
+  %undef = fir.undefined !fir.complex<8>
+  %insert_real = fir.insert_value %undef, %real, %c0 : (!fir.complex<8>, f64, index) -> !fir.complex<8>
```
Comment (schweitzpgi) [inline flang/test/Fir/undo-complex-pattern.fir:92]: Should this test check that the insert_value ops are still present as well?

**Review by schweitzpgi — APPROVED** [review]
